src/btree/bt_cursor.c

```diff
@@ -57,13 +57,36 @@ __cursor_state_restore(WT_CURSOR *cursor, WT_CURFILE_STATE *state)
 static inline bool
 __cursor_page_pinned(WT_CURSOR_BTREE *cbt, bool eviction_ok)
 {
+	WT_SESSION_IMPL *session;
+
+	session = (WT_SESSION_IMPL *)cbt->iface.session;
+
+	if (!F_ISSET(cbt, WT_CBT_ACTIVE))
+		return (false);
+
+	/*
+	 * If we get here during an update with a page lacking history, we'd
+	 * better be able to evict it.
+	 */
+	WT_ASSERT(session, cbt->ref->state != WT_REF_LIMBO ||
+	    !F_ISSET(&session->txn, WT_TXN_UPDATE) || eviction_ok);
```

Comment:
@sueloverso, @keithbostic this is the assertion that fired.  In other words, we get here with a "limbo" page trying to do a `WT_CURSOR::remove` operation and `eviction_ok == false`.

This relates to WT-3211 and WT-3776.  We have to evict in order to get a page that remove can update.  Will that violate the recent bug fix?

Reply:
The original problem was caused by a cursor-remove call positioned to an already removed item, that is, it's possible:

* to have a page populated entirely by deleted items,
* where there's a cursor positioned on one of those items,
* call WT_CURSOR.remove on the item with WT_CURSTD_OVERWRITE set,
* the page is evicted,
* which results in a discarded page (because all the items are deleted),
* the cursor remove then re-instantiates an empty page to do its remove,
* the remove fails, setting the return value to not-found,
* the WT_CURSTD_OVERWRITE configuration resets the return value to 0,
* success implies the cursor should return positioned, 
* there's nothing to position the cursor to,
* core dump.

If there's a chance that a WT_REF_LIMBO page can be full of deleted items during the eviction, there's a problem. Can that happen? I took a look at the branch, but wasn't sure of the answer to that question.

If that can't happen, I think we're OK, and WT_REF_LIMBO can figure into the question of whether or not eviction is permitted.

If that can happen, the alternative I thought about in WT-3776 was to always allow eviction in the cursor-remove path, and define "the cursor is positioned" as "the cursor's key remains set", which scares me some. I don't think there's a lot of functionality at risk here that a user would care about, but it's the kind of edge case we'll get caught on a couple of years from now.

We could also fix this problem by not allowing deletion of a page during a cursor-remove eviction, that is, we won't create an empty page if we're evicting on behalf of cursor-remove and WT_CURSTD_OVERWRITE is configured. It's a nasty little side-case, but I think we could drill the hole without getting too ugly.

Reply:
I've been running the CONFIG and have not yet reproduced the failure.

Do we ever write out a page full of deleted items? If we can then a LIMBO page can be full of deleted items, otherwise it cannot and we're OK.  A LIMBO page is read-only and instantiated from what we read from disk. However `positioned` would be true after a `search`.

Reply:
(Update - it took 100 iterations to reproduce so we can test a fix.)

Reply:
Once an item is deleted, it won't ever be written out, so I don't believe it's possible for row-store to write out a page full of deleted items.